code/game/machinery/atm_ret_field.dm

```diff
@@ -0,0 +1,246 @@
+	var/reboot_delay_min = 50
+	var/reboot_delay_max = 75
+
+	var/hatch_open = FALSE
+	var/wires_intact = TRUE
+	var/list/areas_added
+	var/field_type = /obj/structure/atmospheric_retention_field
+	circuit = /obj/item/circuitboard/arf_generator
+
+/obj/machinery/atmospheric_field_generator/impassable
+	desc = "An older model of ARF-G that generates an impassable retention field. Works just as well as the modern variety, but is slightly more energy-efficient.<br><br>Note: prolonged immersion in active atmospheric retention fields may have negative long-term health consequences."
+	active_power_usage = 2000
+	field_type = /obj/structure/atmospheric_retention_field/impassable
+
+/obj/machinery/atmospheric_field_generator/perma
+	name = "static atmospheric retention field generator"
+	desc = "A floor-mounted piece of equipment that generates an atmosphere-retaining energy field when powered and activated. This model is designed to always be active, though the field will still drop from loss of power or electromagnetic interference.<br><br>Note: prolonged immersion in active atmospheric retention fields may have negative long-term health consequences."
+	alwaysactive = TRUE
+	active_power_usage = 2000
+
+/obj/machinery/atmospheric_field_generator/perma/impassable
+	desc = "An older model of ARF-G that generates an impassable retention field whilst powered. This model's field is quite power-efficient, but will still drop from loss of power or electromagnetic interference.<br><br>Note: prolonged immersion in active atmospheric retention fields may have negative long-term health consequences."
+	active_power_usage = 1500
+	field_type = /obj/structure/atmospheric_retention_field/impassable
+
+/obj/machinery/atmospheric_field_generator/attackby(obj/item/W as obj, mob/user as mob)
+	if(W.is_crowbar() && isactive)
+		if(!src) return
+		to_chat(user, "<span class='warning'>You can't open the ARF-G whilst it's running!</span>")
+		return
+	if(W.is_crowbar() && !isactive)
+		if(!src) return
+		to_chat(user, "<span class='notice'>You [hatch_open? "close" : "open"] \the [src]'s access hatch.</span>")
+		hatch_open = !hatch_open
+		update_icon()
+		if(alwaysactive && wires_intact)
+			generate_field()
+		return
+	if(hatch_open && W.is_multitool())
+		if(!src) return
+		to_chat(user, "<span class='notice'>You toggle \the [src]'s activation behavior to [alwaysactive? "emergency" : "always-on"].</span>")
+		alwaysactive = !alwaysactive
+		update_icon()
+		return
+	if(hatch_open && W.is_wirecutter())
+		if(!src) return
+		to_chat(user, "<span class='warning'>You [wires_intact? "cut" : "mend"] \the [src]'s wires!</span>")
+		wires_intact = !wires_intact
+		update_icon()
+		return
+	if(hatch_open && istype(W,/obj/item/weldingtool))
+		if(!src) return
+		var/obj/item/weldingtool/WT = W
+		if(!WT.isOn())
+			return
+		if(!WT.remove_fuel(0,user))
+			to_chat(user, "<span class='warning'>You need more fuel to complete this task.</span>")
+			return
```

Comment:
Is this not covered by `isOn()` and the actual `remove_fuel()` call below?

Reply:
This is a bit old so I think it predates the remove_fuel rework to have that notification hooked into the base proc itself.

Might need to rework that chunk a bit in general.